SimG4Core/Application/interface/OscarMTMasterThread.h

```diff
@@ -52,23 +58,26 @@ class OscarMTMasterThread {
   std::thread m_masterThread;
 
   // ES products needed for Geant4 initialization
-  mutable edm::ESWatcher<IdealGeometryRecord> idealGeomRcdWatcher_;
-  mutable edm::ESWatcher<IdealMagneticFieldRecord> idealMagRcdWatcher_;
-  mutable const DDCompactView* m_pDD;
-  mutable const cms::DDCompactView* m_pDD4hep;
-  mutable const HepPDT::ParticleDataTable* m_pTable;
-
+  mutable const DDCompactView* m_pDDD = nullptr;
+  mutable const cms::DDCompactView* m_pDD4Hep = nullptr;
+  mutable const HepPDT::ParticleDataTable* m_pTable = nullptr;
+  mutable edm::ESGetToken<DDCompactView, IdealGeometryRecord> m_DDD;
+  mutable edm::ESGetToken<cms::DDCompactView, IdealGeometryRecord> m_DD4Hep;
+  mutable edm::ESGetToken<HepPDT::ParticleDataTable, PDTRecord> m_PDT;
+
+  // status flags
   mutable std::mutex m_protectMutex;
   mutable std::mutex m_threadMutex;
   mutable std::condition_variable m_notifyMasterCv;
   mutable std::condition_variable m_notifyMainCv;
 
   mutable ThreadState m_masterThreadState;
 
-  mutable bool m_masterCanProceed;
-  mutable bool m_mainCanProceed;
-  mutable bool m_firstRun;
-  mutable bool m_stopped;
+  mutable bool m_hasToken = false;
+  mutable bool m_masterCanProceed = false;
+  mutable bool m_mainCanProceed = false;
+  mutable bool m_firstRun = true;
```

Comment:
This no longer does anything. It is set here and in the constructor and set to `false` later but the value is never read and not used for any conditional logic. Previously `m_firstRun` and `idealGeomRcdWatcher_` worked together to guarantee that the geometry only was changed the first time and never again.

Reply:
should we still protect for second use by one flag or it is really meanless?

Reply:
It turned out that these flags are needed. May be some optimization may help to remove one but when I take out two the code stuck forever, so I am in favor have them as it is.